File: jvm/periph-java/src/main/java/it/uhde/periph/chips/humidity/DHT11Full.java
package it.uhde.periph.chips.humidity;

import it.uhde.periph.transport.Transport;
import java.io.IOException;

/**
 * DHT11 -- temperature and humidity sensor (full driver).
 *
 * <p>Extends DHT11Minimal with readTemperature(), readHumidity(),
 * readRetry(), and readRaw() methods.
 */
public class DHT11Full extends DHT11Minimal {

    public DHT11Full(Transport transport) {
        super(transport);
    }

    public double readTemperature() throws IOException {
        return read()[0];
    }

    public double readHumidity() throws IOException {
        return read()[1];
    }

    public double[] readRetry(int maxRetries) throws IOException {
        for (int i = 0; i < maxRetries; i++) {
            try {
                return read();
            } catch (IOException e) {
                // continue
            }
        }
        throw new IOException("all retries exhausted");
    }

    public byte[] readRaw() throws IOException {
        byte[] frame = transport.read(5);

        int humInt  = frame[0] & 0xFF;
        int humDec  = frame[1] & 0xFF;
        int tempInt = frame[2] & 0xFF;
        int tempDec = frame[3] & 0xFF;
        int checksum = frame[4] & 0xFF;

        if ((humInt + humDec + tempInt + tempDec) & 0xFF != checksum) {
            throw new IOException("checksum mismatch");
        }

        return frame;
    }
}

File: jvm/periph-kotlin/src/main/kotlin/it/uhde/periph/chips/humidity/Dht11Full.kt
package it.uhde.periph.chips.humidity

import it.uhde.periph.transport.DHTxxTransport
import java.io.IOException

/**
 * DHT11 — combined temperature and humidity sensor (full driver).
 *
 * Extends [Dht11Minimal] with a configurable-retry read, a
 * [readRetry] convenience method, and [readRaw] access to the
 * unprocessed 5-byte frame.
 */
class Dht11Full @JvmOverloads constructor(
    transport: DHTxxTransport,
    private val maxRetries: Int = 3
) : Dht11Minimal(transport) {

    /**
     * Read both values, retrying on checksum error.
     *
     * @param maxRetries Maximum number of read attempts. If 0, the constructor default is used.
     * @return Pair of (temperature_C, humidity_RH).
     * @throws IOException    on transport error.
     * @throws Dht11Minimal.Dht11Exception if all attempts fail with a checksum error.
     */
    @Throws(IOException::class, Dht11Minimal.Dht11Exception::class)
    fun readRetry(maxRetries: Int = 0): Pair<Double, Double> {
        val n = if (maxRetries > 0) maxRetries else this.maxRetries
        var last: Dht11Minimal.Dht11Exception? = null
        for (i in 0 until n) {
            try {
                return read()
            } catch (e: Dht11Minimal.Dht11Exception) {
                last = e
            }
        }
        throw Dht11Minimal.Dht11Exception("readRetry exhausted after $n attempts: ${last?.message}")
    }

    /**
     * Read the raw 5-byte frame (after validating the checksum).
     *
     * @return 5-byte frame: [hum_int, hum_dec, temp_int, temp_dec, checksum].
     */
    @Throws(IOException::class, Dht11Minimal.Dht11Exception::class)
    fun readRaw(): ByteArray {
        val frame = transport.read()
        decode(frame)
        return frame
    }
}
